spark/src/main/scala/org/apache/spark/sql/delta/DeltaAnalysis.scala

```diff
@@ -917,42 +923,51 @@ class DeltaAnalysis(session: SparkSession)
       attr: NamedExpression,
       targetAttr: NamedExpression,
       tblName: String,
-      allowTypeWidening: Boolean): NamedExpression = {
+      typeWideningMode: TypeWideningMode): NamedExpression = {
     val expr = (attr.dataType, targetAttr.dataType) match {
       case (s, t) if s == t =>
         attr
       case (s: StructType, t: StructType) if s != t =>
-        addCastsToStructs(tblName, attr, s, t, allowTypeWidening)
+        addCastsToStructs(tblName, attr, s, t, typeWideningMode)
       case (ArrayType(s: StructType, sNull: Boolean), ArrayType(t: StructType, tNull: Boolean))
-          if s != t && sNull == tNull =>
-        addCastsToArrayStructs(tblName, attr, s, t, sNull, allowTypeWidening)
+        if s != t && sNull == tNull =>
+        addCastsToArrayStructs(tblName, attr, s, t, sNull, typeWideningMode)
       case (s: AtomicType, t: AtomicType)
-        if allowTypeWidening && TypeWidening.isTypeChangeSupportedForSchemaEvolution(t, s) =>
+        if typeWideningMode.shouldWidenType(fromType = t, toType = s) =>
         // Keep the type from the query, the target schema will be updated to widen the existing
         // type to match it.
         attr
       case (s: MapType, t: MapType)
-        if !DataType.equalsStructurally(s, t, ignoreNullability = true) || allowTypeWidening =>
-        // only trigger addCastsToMaps if exists differences like extra fields, renaming
-        // Or allowTypeWidening is enabled
-        addCastsToMaps(tblName, attr, s, t, allowTypeWidening)
+        if !DataType.equalsStructurally(s, t, ignoreNullability = true) =>
+        // only trigger addCastsToMaps if exists differences like extra fields, renaming or type
+        // differences.
+        addCastsToMaps(tblName, attr, s, t, typeWideningMode)
       case _ =>
         getCastFunction(attr, targetAttr.dataType, targetAttr.name)
     }
     Alias(expr, targetAttr.name)(explicitMetadata = Option(targetAttr.metadata))
   }
 
   /**
-   * Whether inserting values that have a wider type than the table has is allowed. In that case,
-   * values are not downcasted to the current table type and the table schema is updated instead to
-   * use the wider type.
+   * Returns a mapping of (fromType, toType) to Boolean indicating whether `fromType` is eligible to
+   * be automatically widened to `toType` when ingesting data. If it is, the table schema is updated
+   * to `toType` before ingestion and values are written using their origin `toType` type.
+   * Otherwise, the table type `fromType` is retained and values are downcasted on write.
```

Comment:
Need to update this comment now that it actually returns a mode instance, not a function anymore.

Reply:
Done